src/tutorials/ex3.c

```diff
@@ -0,0 +1,186 @@
+
+static char help[] = "Solves a tridiagonal system with lower bound specified as an inequality constraint.\n\
+Sovles finite differences discretizatio of:\n\
+-u''(x) = -15,  x in [0,1]\n\
+u(0) = u(1) = 0\n\
+s.t. u(x) >= sin(4*pi*x -pi/6)/2 -2\n\
+Based on ex1.\n\
+Input parameters include:\n\
+  -n <mesh_n> : number of mesh points\n\
+  -sol        : view and draw the solution vector\n\
+  -draw_puase : number of seconds to pause, -1 implies until user input, see PetscDrawSetPause()\n";
+
+/*
+* Include "permonqps.h" so that we can use QPS solvers.  Note that this file
+* automatically includes:
+*   petscsys.h   - base PERMON routines
+*   permonvec.h  - Vectors
+*   permonmat.h  - Matrices
+*   permonqppf.h - Projection Factory
+*   permonqpc.h  - Quadratic Programming Constraints
+*   permonqp.h   - Quadratic Programming objects and transformations
+*   petsctao.h   - Toolkit for Advanced Optimization solvers
+*/
+#include <permonqps.h>
+#include <petscdraw.h>
+#include <math.h>
+#define PI 3.14159265
+
+/* Draw vector */
+PetscErrorCode viewDraw(Vec x) {
+  PetscViewer    v1;
+  PetscDraw      draw;
+  PetscErrorCode ierr;
+
+  PetscFunctionBeginUser;
+  ierr = PetscViewerDrawOpen(PETSC_COMM_WORLD,0,"",80,380,400,160,&v1);CHKERRQ(ierr);
+  ierr = PetscViewerDrawGetDraw(v1,0,&draw);CHKERRQ(ierr);
+  ierr = PetscDrawSetDoubleBuffer(draw);CHKERRQ(ierr);
+  ierr = PetscDrawSetFromOptions(draw);CHKERRQ(ierr);
+  ierr = VecView(x,v1);CHKERRQ(ierr);
+  ierr = PetscViewerDestroy(&v1);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+/* Lower bound (obstacle) function */
+PetscReal fobst(PetscInt i,PetscInt n) {
+  PetscReal h = 1./(n-1);
+  return sin(4*PI*i*h-PI/6.)/2 -2; 
+}
+
+#undef __FUNCT__
+#define __FUNCT__ "main"
```

Comment:
Keep consistency of using or not using preprocessor directives associated with functions.

Reply:
These should be removed everywhere. PETSc has not used them for a long time already.